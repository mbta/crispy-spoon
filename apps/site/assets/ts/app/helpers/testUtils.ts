/* eslint-disable */
import toJson, { Json } from "enzyme-to-json";
import { ReactWrapper } from "enzyme";
import {
  IndexedRouteStop,
  StopTree
} from "../../schedule/components/__schedule";

export const testRouteStopListFromStopTree = (
  tree: StopTree
): IndexedRouteStop[] =>
  Object.values(tree.byId).map((node, index) => ({
    ...node.value,
    routeIndex: index
  }));

export const createReactRoot = (): void => {
  document.body.innerHTML =
    '<div><div id="react-root"><div id="test"></div></div></div>';
};

export const enzymeToJsonWithoutProps = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false,
    map: info => {
      if (!info.props) {
        return info;
      }

      // @ts-ignore is a <style></style> declaration
      if (info.node.type === "style") {
        return info;
      }
      // @ts-ignore is an element in a <noscript></noscript> declaration
      if (info.node.nodeType == undefined) {
        // @ts-ignore
        return info;
      }
      //@ts-ignore is a React component (not HTML element)
      if (info.node.nodeType.toString().includes("function")) {
        // @ts-ignore
        info.props = undefined;
      }

      return info;
    }
  });

export const toJsonEnzyme = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false
  });
