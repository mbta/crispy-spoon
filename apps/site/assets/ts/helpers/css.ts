import {
  isABusRoute,
  isACommuterRailRoute,
  isAGreenLineRoute,
  isASilverLineRoute
} from "../models/route";
import { Route } from "../__v3api";

export const routeToCSSClass = (route: string): string =>
  route.toLowerCase().replace(/[ _]/g, "-");

export const busClass = (route: Route): string =>
  isABusRoute(route) && !isASilverLineRoute(route) ? "bus-route-sign" : "";

export const routeToModeName = (route: Route): string => {
  if (isACommuterRailRoute(route)) return "commuter-rail";
  if (route.type === 4) return "ferry";
  if (route.id === "Red" || route.id === "Mattapan") return "red-line";
  if (route.id === "Orange") return "orange-line";
  if (route.id === "Blue") return "blue-line";
  if (isAGreenLineRoute(route)) return "green-line";
  if (isASilverLineRoute(route)) return "silver-line";
  if (isABusRoute(route)) return "bus";
  return "unknown";
};

export const routeBgClass = (route: Route): string =>
  `u-bg--${routeToModeName(route)}`;
