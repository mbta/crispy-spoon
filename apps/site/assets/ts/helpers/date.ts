// Returns a date in the form of "August 7, 2019"
export const formattedDate = (unformatted: string): string => {
  const [year, month, day] = unformatted
    .split(/-/)
    .map(part => Number.parseInt(part, 10));

  // Remember that months in JS are 0-indexed for some reason, hence "month - 1".
  const parsedDate = new Date(year, month - 1, day);
  return parsedDate.toLocaleDateString("en-US", {
    year: "numeric",
    month: "long",
    day: "numeric",
    timeZone: "UTC"
  });
};

// Returns a date in the form of "Aug 7"
export const shortDate = (date: Date): string =>
  date.toLocaleDateString("en-US", {
    month: "short",
    day: "numeric",
    timeZone: "UTC"
  });

export default formattedDate;
