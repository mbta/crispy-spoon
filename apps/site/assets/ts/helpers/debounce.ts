// Debounce is meant to be used with event listeners. Use it to delay a call
// until after a burst of events, such as keyboard input.
//
//   usage:
//
//     const callback = (ev: Event) => {}
//     document.addEventListener("keyup", debounce(callback, 250))
//

const debounce = (callback: Function, delay: number): EventListener => {
  let timer: number | undefined;

  return (...args) => {
    window.clearTimeout(timer);
    timer = window.setTimeout(() => callback(...args), delay);
  };
};

export const debouncePromise = (
  callback: Function,
  delay: number
): Function => {
  let timerId: number | undefined;

  return function debounced(...args: unknown[]) {
    if (timerId) {
      window.clearTimeout(timerId);
    }

    return new Promise(resolve => {
      timerId = window.setTimeout(() => resolve(callback(...args)), delay);
    });
  };
};

export default debounce;
