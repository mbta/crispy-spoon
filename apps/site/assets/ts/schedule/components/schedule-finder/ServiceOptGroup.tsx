import React, { ReactElement } from "react";
import ServiceOption from "./ServiceOption";
import HolidayServiceOption from "./HolidayServiceOption";
import {
  ServiceOptGroupName,
  ServiceByOptGroup
} from "../../../helpers/service";

interface Props {
  group: ServiceOptGroupName;
  label: string;
  services: ServiceByOptGroup[];
  multipleWeekdays: boolean;
}

const ServiceOptGroup = ({
  group,
  label,
  services,
  multipleWeekdays
}: Props): ReactElement<HTMLElement> | null => (
  <optgroup key={group} label={label}>
    {services.map((service: ServiceByOptGroup) =>
      group === "holiday" ? (
        <HolidayServiceOption
          key={service.service.id + service.servicePeriod}
          service={service.service}
          servicePeriod={service.servicePeriod}
        />
      ) : (
        <ServiceOption
          key={service.service.id + service.servicePeriod}
          service={service.service}
          group={group}
          servicePeriod={service.servicePeriod}
          multipleWeekdays={multipleWeekdays}
        />
      )
    )}
  </optgroup>
);
export default ServiceOptGroup;
