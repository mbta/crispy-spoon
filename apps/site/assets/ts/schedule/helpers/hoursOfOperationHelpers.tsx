import React, { ReactElement } from "react";
import renderFa from "../../helpers/render-fa";
import { SchedulePDF } from "../components/__schedule";

const pdfLink = (pdf: SchedulePDF | null): ReactElement<HTMLElement> | null => {
  if (!pdf) {
    return null;
  }
  return (
    <div className="fs-16 pt-16">
      <a
        key={pdf.url}
        href={pdf.url}
        rel="noopener noreferrer"
        className="c-call-to-action"
        target="_blank"
      >
        {`Open full schedule and map PDF`}{" "}
        {renderFa("fa-arrow-up-right-from-square", "")}
      </a>
    </div>
  );
};

export default pdfLink;
