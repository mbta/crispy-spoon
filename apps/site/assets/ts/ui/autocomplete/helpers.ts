import { omit } from "lodash";
import { OnSubmitParams } from "@algolia/autocomplete-core";
import { OnStateChangeProps } from "@algolia/autocomplete-js";
import {
  ContentItem,
  Item,
  LocationItem,
  PopularItem,
  RouteItem,
  StopItem
} from "./__autocomplete";
import { isLGDown } from "../../helpers/media-breakpoints";

export function isStopItem(x: Item): x is StopItem {
  return Object.keys(x).includes("stop");
}

export function isRouteItem(x: Item): x is RouteItem {
  return Object.keys(x).includes("route");
}

export function isContentItem(x: Item): x is ContentItem {
  return Object.keys(x).includes("_content_type");
}

export const getTitleAttribute = (item: Item): string[] => {
  if (isStopItem(item)) {
    return ["stop", "name"];
  }
  if (isRouteItem(item)) {
    return ["route", "name"];
  }

  // @ts-ignore
  // eslint-disable-next-line no-underscore-dangle
  return Object.keys(item._highlightResult);
};

const navStateChange: (props: OnStateChangeProps<Item>) => void = ({
  state
}) => {
  // grey out the page and disable scrolling when search is open
  if (isLGDown()) {
    if (state.isOpen) {
      document.documentElement.dataset.navOpen = "true";
    } else {
      delete document.documentElement.dataset.navOpen;
    }
  }
};

export const STATE_CHANGE_HANDLERS: Record<
  string,
  (props: OnStateChangeProps<Item>) => void
> = {
  nav: navStateChange
};

export const SUBMIT_HANDLERS: Record<
  string,
  (props: OnSubmitParams<Item>) => void
> = {
  to_search_page({ state }) {
    window.Turbolinks.visit(`/search?query=${state.query}`);
  }
};

export type WithUrls<T> = T & { urls: Record<string, string> };

// The backend returns all possible URLs, use urlType to get the desired one
export const itemWithUrl = (
  initialItem: WithUrls<LocationItem | PopularItem>,
  urlType: string
): Omit<typeof initialItem, "urls"> => {
  const item = omit(initialItem, "urls");
  item.url = initialItem.urls[urlType];
  return item;
};
