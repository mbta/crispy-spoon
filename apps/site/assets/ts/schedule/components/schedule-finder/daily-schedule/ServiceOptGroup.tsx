import React, { ReactElement } from "react";
import { format } from "date-fns";
import { Service } from "../../../../__v3api";
import { shortDate, stringToDateObject } from "../../../../helpers/date";
import {
  ServiceGroupNames,
  serviceDays,
  dedupeServices,
  isInCurrentService
} from "../../../../helpers/service";

interface Props {
  label: string;
  services: Service[];
  multipleWeekdays: boolean;
  todayServiceId: string;
  nowDate?: Date;
}

const ServiceOptGroup = ({
  label,
  services,
  multipleWeekdays,
  todayServiceId,
  nowDate = new Date()
}: Props): ReactElement<HTMLElement> | null =>
  services.length === 0 ? null : (
    <optgroup label={label}>
      {dedupeServices(services).map(service => {
        const isMultipleWeekday =
          multipleWeekdays &&
          service.type === "weekday" &&
          service.typicality !== "holiday_service";

        const startDate = stringToDateObject(service.start_date);
        const endDate = stringToDateObject(service.end_date);
        let optionText = "";

        if (
          service.typicality === "unplanned_disruption" ||
          service.typicality === "planned_disruption"
        ) {
          optionText = service.description;
        } else if (
          service.typicality === "holiday_service" &&
          service.added_dates &&
          service.added_dates_notes
        ) {
          const addedDate = service.added_dates[0];
          const addedNote = service.added_dates_notes[addedDate];
          const addedNoteString = addedNote ? `${addedNote}, ` : "";
          optionText = `${addedNoteString}${shortDate(
            stringToDateObject(addedDate)
          )}`;
        } else if (label === ServiceGroupNames.OTHER) {
          optionText = isMultipleWeekday
            ? `${serviceDays(service)} schedule`
            : service.description;
          optionText += `, ${shortDate(startDate)} to ${shortDate(endDate)}`;
        } else {
          optionText = isMultipleWeekday
            ? `${serviceDays(service)} schedule`
            : service.description;
          if (service.rating_description) {
            optionText += `, ${service.rating_description}`;
          }
        }

        if (service.id === todayServiceId) {
          // if it's the only service this rating, this service might not
          // technically be now - adjust text for that
          if (isInCurrentService(service, nowDate)) {
            optionText += " (now)";
          } else {
            optionText += ` (Starting ${format(startDate, "MMMM d, y")})`;
          }
        }

        return (
          <option
            key={service.id}
            value={service.id}
            // the <optgroup> is not reliably conveyed to assistive tech, so add extra label here
            aria-label={`${label} ${optionText}`}
          >
            {optionText}
          </option>
        );
      })}
    </optgroup>
  );

export default ServiceOptGroup;
