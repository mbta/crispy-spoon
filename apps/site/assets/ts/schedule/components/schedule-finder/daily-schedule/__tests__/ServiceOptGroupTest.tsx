import { mount } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";
import { createReactRoot } from "../../../../../app/helpers/testUtils";
import { services } from "../../../../../helpers/__tests__/service-test";
import { Service } from "../../../../../__v3api";
import ServiceOptGroup from "../ServiceOptGroup";

describe("ServiceOptGroup", () => {
  it("renders", () => {
    createReactRoot();
    const tree = renderer
      .create(
        <ServiceOptGroup
          label={"Group label"}
          services={services}
          multipleWeekdays={false}
          todayServiceId=""
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("returns null if no services", () => {
    createReactRoot();
    const tree = renderer.create(
      <ServiceOptGroup
        label={""}
        services={[]}
        multipleWeekdays={false}
        todayServiceId=""
      />
    );
    expect(tree.toJSON()).toBeNull();
  });

  const makeSimpleService = (
    [start_date, end_date]: [string, string],
    [rating_start_date, rating_end_date]: [string, string],
    name: string,
    id: string
  ): Service => ({
    valid_days: [1, 2, 3, 4, 5],
    typicality: "typical_service",
    type: "weekday",
    start_date,
    removed_dates_notes: {},
    removed_dates: [],
    name,
    id,
    end_date,
    description: `${name} schedule`,
    added_dates_notes: {},
    added_dates: [],
    rating_start_date,
    rating_end_date,
    rating_description: "Test Rating"
  });

  const servicesList: Service[] = [
    makeSimpleService(
      ["2019-04-01", "2019-05-01"],
      ["2019-04-01", "2019-08-01"],
      "Spring",
      "s"
    ),
    makeSimpleService(
      ["2019-06-01", "2019-07-01"],
      ["2019-04-01", "2019-08-01"],
      "First Summer",
      "fs"
    ),
    makeSimpleService(
      ["2019-08-06", "2019-09-01"],
      ["2019-08-01", "2019-10-01"],
      "Second Summer",
      "ss"
    )
  ];

  it("adds (now) to today's service based on todayServiceId prop", () => {
    const wrapper = mount(
      <ServiceOptGroup
        label={"Test services"}
        services={servicesList.concat(
          makeSimpleService(
            ["2019-09-06", "2019-10-01"],
            ["2019-09-01", "2019-11-01"],
            "Current Summer",
            "n"
          )
        )}
        multipleWeekdays={false}
        todayServiceId="n"
        nowDate={new Date("2019-09-07")}
      />
    );
    expect(wrapper.text()).toContain(
      "Current Summer schedule, Test Rating (now)"
    );
  });

  it("adds (starting date) to today's service based on todayServiceId prop if service hasn't yet started", () => {
    const wrapper = mount(
      <ServiceOptGroup
        label={"Test services"}
        services={servicesList}
        multipleWeekdays={false}
        todayServiceId="fs"
      />
    );
    expect(wrapper.text()).toContain(
      "First Summer schedule, Test Rating (Starting June 1, 2019)"
    );
  });

  it("does not mark any service as (now) if no current services", () => {
    const wrapper = mount(
      <ServiceOptGroup
        label={"Test services"}
        services={servicesList}
        multipleWeekdays={false}
        todayServiceId=""
      />
    );
    expect(wrapper.text()).not.toContain("(now)");
  });
});
