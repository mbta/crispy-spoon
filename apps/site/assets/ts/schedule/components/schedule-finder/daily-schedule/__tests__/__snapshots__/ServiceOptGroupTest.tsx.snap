// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceOptGroup renders 1`] = `
<optgroup
  label="Group label"
>
  <option
    aria-label="Group label Weekday schedule, Test Rating"
    value="BUS319-O-Wdy-02"
  >
    Weekday schedule, Test Rating
  </option>
  <option
    aria-label="Group label Weekday schedule, Test Rating"
    value="BUS319-D-Wdy-02"
  >
    Weekday schedule, Test Rating
  </option>
  <option
    aria-label="Group label Saturday schedule, Test Rating"
    value="BUS319-P-Sa-02"
  >
    Saturday schedule, Test Rating
  </option>
  <option
    aria-label="Group label Sunday schedule, Test Rating"
    value="BUS319-Q-Su-02"
  >
    Sunday schedule, Test Rating
  </option>
  <option
    aria-label="Group label Bastille Day Eve, Jul 13"
    value="BastilleDayEve"
  >
    Bastille Day Eve, Jul 13
  </option>
  <option
    aria-label="Group label Bastille Day, Jul 14"
    value="BastilleDay"
  >
    Bastille Day, Jul 14
  </option>
  <option
    aria-label="Group label Jul 15"
    value="2019-07-15"
  >
    Jul 15
  </option>
  <option
    aria-label="Group label Storm (reduced service)"
    value="BUS319-storm"
  >
    Storm (reduced service)
  </option>
  <option
    aria-label="Group label Another service, Test Future Rating"
    value="weekday2020"
  >
    Another service, Test Future Rating
  </option>
  <option
    aria-label="Group label Ferry service"
    value="weekday2019"
  >
    Ferry service
  </option>
  <option
    aria-label="Group label CR service, Test Future Service Incomplete Rating"
    value="weekday2020-2"
  >
    CR service, Test Future Service Incomplete Rating
  </option>
</optgroup>
`;
